src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_black_knight.cpp

```diff
@@ -15,337 +15,470 @@
+            }
+        }
+
+        void DoExplode()
         {
-            summons.DespawnAll();
-            me->SetDisplayId(me->GetNativeDisplayId());
-            me->ClearUnitState(UNIT_STATE_ROOT | UNIT_STATE_STUNNED);
+            // We trigger the ghoul explode spell casted on us by Black Knight
+            // to get immunity to crowd control forms and we get correct growth aura
+            // the spell script of Ghoul Explode functions normally from here on
+            if (me->HasAura(SPELL_GHOUL_EXPLODE))
+                return;
+            DoCast(me, SPELL_GHOUL_EXPLODE, true);
+            // wtf?
+            if (me->GetEntry() == NPC_RISEN_ARELAS || me->GetEntry() == NPC_RISEN_JAEREN)
+                Talk(urand(0, 2));
+        }
 
-            Initialize();
+        void DamageTaken(Unit* /*attacker*/, uint32& /*damage*/) override
+        {
+            if (HealthBelowPct(31) && !doExplode)
+            {
+                // Ghouls cast explode when they reach 30% health or if Black Knight dies
+                // or if Black Knight forces them to explode
+                doExplode = true;
+                DoExplode();
+            }
         }
 
-        void JustSummoned(Creature* summon) override
+        void JustEngagedWith(Unit* /*who*/) override
         {
-            summons.Summon(summon);
-            summon->AI()->AttackStart(me->GetVictim());
+            events.ScheduleEvent(EVENT_CLAW, 3000);
+            events.ScheduleEvent(EVENT_LEAP, 2000);
+            DoZoneInCombat();
         }
 
-        void SummonedCreatureDespawn(Creature* summon) override
+        void JustDied(Unit* /*killer*/) override
         {
-            summons.Despawn(summon);
+            me->DespawnOrUnsummon(1000);
         }
 
-        void UpdateAI(uint32 uiDiff) override
+        void KilledUnit(Unit* who) override
         {
-            //Return since we have no target
-            if (!UpdateVictim())
+            if (who == me)
                 return;
-
-            if (bEventInProgress)
+            if (Creature* knight = instance->GetCreature(DATA_BLACK_KNIGHT))
             {
-                if (uiResurrectTimer <= uiDiff)
-                {
-                    me->SetFullHealth();
-                    DoCast(me, SPELL_BLACK_KNIGHT_RES, true);
-                    uiPhase++;
-                    uiResurrectTimer = 4000;
-                    bEventInProgress = false;
-                    me->ClearUnitState(UNIT_STATE_ROOT | UNIT_STATE_STUNNED);
-                } else uiResurrectTimer -= uiDiff;
+                if (!knight->HasAura(SPELL_BLACK_KNIGHT_DIE))
+                    knight->AI()->KilledUnit(who);
             }
+        }
+
+        void UpdateAI(uint32 uiDiff) override
+        {
+            events.Update(uiDiff);
 
-            switch (uiPhase)
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            if (!UpdateVictim())
```

Comment:
why is this after the eventmap call/casting check? should these not evade while casting?